client/exec.go

```diff
@@ -0,0 +1,305 @@
+			go func() {
+				wsutil.WebsocketSendStream(conn, args.Stdin, -1)
+				<-wsutil.WebsocketRecvStream(args.Stdout, conn)
+				conn.Close()
+				if args.DataDone != nil {
+					close(args.DataDone)
+				}
+			}()
+		} else {
+			if args.DataDone != nil {
+				close(args.DataDone)
+			}
+		}
+	} else {
+		// Handle non-terminal executions
+		dones := map[string]chan bool{}
+		conns := []*websocket.Conn{}
+
+		// Handle stdin and stdout
+		if fds["io"] != "" {
+			conn, err := client.getChangeWebsocket(changeID, fds["io"])
+			if err != nil {
+				return "", err
+			}
+			conns = append(conns, conn)
+			dones["stdin"] = wsutil.WebsocketSendStream(conn, args.Stdin, -1)
+			dones["stdout"] = wsutil.WebsocketRecvStream(args.Stdout, conn)
+		}
+
+		// Handle stderr separately if needed
+		if opts.Stderr && fds["stderr"] != "" {
+			conn, err := client.getChangeWebsocket(changeID, fds["stderr"])
+			if err != nil {
+				return "", err
+			}
+			conns = append(conns, conn)
+			dones["stderr"] = wsutil.WebsocketRecvStream(args.Stderr, conn)
+		}
+
+		// Wait for everything to be done
+		go func() {
+			for name, done := range dones {
+				// Skip stdin, dealing with it separately below
+				if name == "stdin" {
+					continue
+				}
+				<-done
+			}
+
+			if fds["io"] != "" {
+				if args.Stdin != nil {
+					args.Stdin.Close()
+				}
+
+				// Empty the stdin channel but don't block on it as
+				// stdin may be stuck in Read()
+				go func() {
+					<-dones["stdin"]
+				}()
+			}
+
+			for _, conn := range conns {
+				conn.Close()
+			}
+
+			if args.DataDone != nil {
+				close(args.DataDone)
+			}
+		}()
+	}
+
+	return changeID, nil
+}
+
+// getChangeWebsocket creates a websocket connection for the given change ID
+// and websocket ID combination.
+func (client *Client) getChangeWebsocket(changeID, websocketID string) (*websocket.Conn, error) {
+	// Set up a new websocket dialer based on the HTTP client
+	httpClient := client.doer.(*http.Client)
+	httpTransport := httpClient.Transport.(*http.Transport)
+	dialer := websocket.Dialer{
+		NetDial:          httpTransport.Dial,
+		Proxy:            httpTransport.Proxy,
+		TLSClientConfig:  httpTransport.TLSClientConfig,
+		HandshakeTimeout: 5 * time.Second,
+	}
+
+	// Establish the connection
+	url := fmt.Sprintf("ws://localhost/v1/changes/%s/websocket?id=%s", changeID, url.QueryEscape(websocketID))
+	conn, _, err := dialer.Dial(url, nil)
+	if err != nil {
+		return nil, err
+	}
+	return conn, err
+}
+
+// WebsocketWriter is a websocket writer interface that can write a websocket
+// or a value as JSON, for example, for sending commands to an executing
+// program's "control" websocket.
+type WebsocketWriter interface {
+	WriteMessage(messageType int, data []byte) error
+	WriteJSON(v interface{}) error
+}
+
+type execCommand struct {
+	Command string            `json:"command"`
+	Args    map[string]string `json:"args,omitempty"`
+	Signal  int               `json:"signal,omitempty"`
```

Comment:
Hmmm, should this be a string like `"SIGHUP"` instead of the signal number? Seems better for an API, and signal names are more cross-platform than signal numbers (which are different on macOS, for example). Names would make it unambiguous, and is also what I've suggested for the new [`/v1/signals` API in JU011](https://docs.google.com/document/d/1d6-h3UAt2VPUSvlkVF30l8iuDW8raRNkHbp5M6NUo1A/edit#heading=h.5psp0nwlh2a2).

Reply:
Changed to a string for the above reasons.